File: CSharp/ExtensibleCommands/ExtensibleCommands/TryCatchFinallyCommand.cs
using System;
using System.Collections.Generic;

namespace ExtensibleCommands
{
    /// <summary>
    /// Executes Core command, and then executes Finally command, 
    /// regardless of the result of the Core command.
    /// Finally command will always be executed (unless unhandled exception is thrown in Core command).
    /// </summary>
    public class TryCatchFinallyCommand : DecoratorCommand
    {
        /// <summary> Recovery command to execute after the Core command is completed with or without error </summary>
        public ICommand FinallyCommand { get; }
        
        /// <summary> List of all child command objects (1st level only) </summary>
        public override IEnumerable<ICommand> Children
        {
            get
            {
                var children = new List<ICommand>(base.Children);
                children.Add(FinallyCommand);
                return children;
            }
        }

        /// <summary> List of all descendant command objects (all levels) </summary>
        public override IEnumerable<ICommand> Descendants
        {
            get
            {
                var descendants = new List<ICommand>(base.Descendants);
                descendants.Add(FinallyCommand);
                descendants.AddRange(FinallyCommand.Descendants);
                return descendants;
            }
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreCommand">Core command</param>
        /// <param name="finallyCommand">Recovery command to execute after the Core command is completed with or without error</param>
        public TryCatchFinallyCommand(ICommand coreCommand, ICommand finallyCommand)
            : this(coreCommand, finallyCommand, "Try-Catch-Finally") { }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="coreCommand">Core command</param>
        /// <param name="finallyCommand">Recovery command to execute after the Core command is completed with or without error</param>
        /// <param name="name">Command name</param>
        public TryCatchFinallyCommand(ICommand coreCommand, ICommand finallyCommand, string name)
            : base(coreCommand, name)
        {
            if (finallyCommand == null)
                throw new Exception(string.Format("Core Command is NULL in TryCatchFinallyCommand {0}", Name));

            FinallyCommand = finallyCommand;
        }

        protected override void Execute()
        {
            // Run core command
            CoreCommand.Run();

            ProcessAbortAndPauseEvents();

            if (CoreCommand.CurrentState == State.Aborted || CurrentState == State.Aborted) return;

            // Remember Core command exception
            if (CoreCommand.CurrentState == State.Failed)
            {
                Exception = CoreCommand.Exception;
            }

            FinallyCommand.Run();
        }

        /// <summary> Set the main command state based on the child commands states </summary>
        protected override void CheckErrors()
        {
            if (CoreCommand.CurrentState == State.Aborted ||
               FinallyCommand.CurrentState == State.Aborted)
            {
                CurrentState = State.Aborted;
            }
            if (FinallyCommand.CurrentState == State.Failed)
            {
                // If Finally command failed, the whole command failed
                CurrentState = State.Failed;
                Exception = FinallyCommand.Exception;
            }
            else if (CoreCommand.CurrentState == State.Failed)
            {
                Exception = CoreCommand.Exception;
                // If Core command failed, the whole command failed
                CurrentState = State.Failed;
            }
            else if (CoreCommand.CurrentState == State.Completed &&
               FinallyCommand.CurrentState == State.Completed)
            {
                CurrentState = State.Completed;
            }
        }
    }
}

File: Java/ExtensibleCommands/src/main/java/org/extensiblecommands/TryCatchFinallyCommand.java
package org.extensiblecommands;

import java.util.ArrayList;

import static java.util.Objects.requireNonNull;

/**
 * Executes Core command, and then executes Finally command,
 * regardless of the result of the Core command.
 * Finally command will always be executed (unless an unhandled exception is thrown in the Core command).
 */
public class TryCatchFinallyCommand extends DecoratorCommand {
    /**
     * Recovery command to execute after the Core command is completed with or without error
     */
    private final Command finallyCommand;

    /**
     * Constructor
     * @param coreCommand       Core command
     * @param finallyCommand    Recovery command to execute after the Core command is completed with or without error
     */
    public TryCatchFinallyCommand(Command coreCommand, Command finallyCommand) {
        this(coreCommand, finallyCommand, "Try-Catch-Finally");
    }

    /**
     * Constructor
     * @param coreCommand       Core command
     * @param finallyCommand    Recovery command to execute after the Core command is completed with or without error
     * @param name              Command name
     */
    public TryCatchFinallyCommand(Command coreCommand, Command finallyCommand, String name) {
        super(coreCommand, name);

        requireNonNull(finallyCommand, String.format("Core Command is NULL in TryCatchFinallyCommand %s", name));

        this.finallyCommand = finallyCommand;
    }

    /**
     * @return          Recovery command to execute after the Core command is completed with or without error
     */
    public final Command getFinallyCommand() {
        return finallyCommand;
    }

    /**
     * @return          List of all child command objects (1st level only)
     */
    @Override
    public Iterable<Command> getChildren() {
        var children = new ArrayList<Command>();
        super.getChildren().forEach(c -> children.add(c));
        children.add(finallyCommand);
        return children;
    }

    /**
     * @return          List of all descendant command objects (all levels)
     */
    @Override
    public Iterable<Command> getDescendants() {
        var descendants = new ArrayList<Command>();
        super.getDescendants().forEach(c -> descendants.add(c));
        descendants.add(finallyCommand);
        finallyCommand.getDescendants().forEach(c -> descendants.add(c));
        return descendants;
    }

    @Override
    protected void execute() throws Exception {
        // Run core command
        coreCommand.run();

        processAbortAndPauseEvents();

        if (coreCommand.getState() == State.Aborted || getState() == State.Aborted)
            return;

        // Remember Core command exception
        if (coreCommand.getState() == State.Failed)
            exception = coreCommand.getException();

        finallyCommand.run();
    }

    /**
     * Set the main command state based on the child commands states
     */
    @Override
    protected void checkErrors() {
        if (coreCommand.getState() == State.Aborted || finallyCommand.getState() == State.Aborted) {
            setState(State.Aborted);
        }
        if (finallyCommand.getState() == State.Failed) {
            // If Finally command failed, the whole command failed
            setState(State.Failed);
            exception = finallyCommand.getException();
        }
        else if (coreCommand.getState() == State.Failed) {
            exception = coreCommand.getException();
            // If Core command failed, the whole command failed
            setState(State.Failed);
        }
        else if (coreCommand.getState() == State.Completed &&
                finallyCommand.getState() == State.Completed) {
            setState(State.Completed);
        }
    }
}
